contracts/external/traders/LimitOrders.sol

```diff
@@ -0,0 +1,583 @@
+            FILE,
+            "Cannot cancel canceled order"
+        );
+        g_status[canceler][orderHash] = OrderStatus.Canceled;
+        emit LogOrderCanceled(canceler, orderHash);
+    }
+
+    /**
+     * Approves an order as long as it is not already approved or canceled.
+     */
+    function approveOrderInternal(
+        address approver,
+        bytes32 orderHash
+    )
+        private
+    {
+        Require.that(
+            g_status[approver][orderHash] == OrderStatus.Null,
+            FILE,
+            "Cannot approve non-null order"
+        );
+        g_status[approver][orderHash] = OrderStatus.Approved;
+        emit LogOrderApproved(approver, orderHash);
+    }
+
+    // ============ Private Helper Functions ============
+
+    /**
+     * Returns the trade cost for some order and inputWei.
+     */
+    function getTradeCostInternal(
+        uint256 inputMarketId,
+        uint256 outputMarketId,
+        Types.Wei memory inputWei,
+        Order memory order
+    )
+        private
+        returns (Types.AssetAmount memory)
+    {
+        Require.that(
+            !inputWei.isZero(),
+            FILE,
+            "InputWei is zero"
+        );
+        Require.that(
+            (order.makerMarket == outputMarketId && order.takerMarket == inputMarketId) ||
+            (order.takerMarket == outputMarketId && order.makerMarket == inputMarketId),
+            FILE,
+            "Market mismatch"
+        );
+        Require.that(
+            inputWei.sign == (order.takerMarket == inputMarketId),
+            FILE,
+            "InputWei sign mismatch"
+        );
+
+        uint256 outputAmount;
+        uint256 fillAmount;
+
+        if (order.takerMarket == inputMarketId) {
+            outputAmount = inputWei.value.getPartial(order.makerAmount, order.takerAmount);
+            fillAmount = outputAmount;
+        } else {
+            assert(order.takerMarket == outputMarketId);
+            outputAmount = inputWei.value.getPartialRoundUp(order.takerAmount, order.makerAmount);
+            fillAmount = inputWei.value;
+        }
+
+        updateFilledAmount(order, fillAmount);
+
+        return Types.AssetAmount({
+            sign: order.takerMarket == outputMarketId,
+            denomination: Types.AssetDenomination.Wei,
+            ref: Types.AssetReference.Delta,
+            value: outputAmount
+        });
+    }
+
+    /**
+     * Increases the stored filled amount (in makerAmount) of the order by fillAmount.
+     */
+    function updateFilledAmount(
+        Order memory order,
+        uint256 fillAmount
+    )
+        private
+    {
+        uint256 newFillAmount = g_filled[order.orderHash].add(fillAmount);
+        Require.that(
+            newFillAmount <= order.makerAmount,
+            FILE,
+            "Cannot overfill order"
+        );
+        g_filled[order.orderHash] = newFillAmount;
+        emit LogOrderTaken(order.orderHash, fillAmount);
```

Comment:
this is kind of a weird place to log an event (buried in 2 helper functions and in a function called "update..")

Reply:
moved up one level